Add support for `com.github.ben-manes.caffeine:caffeine:2.9.3`

linghengqian opened the pull request:
## What does this PR do?

- Fixes #384 .
- Add support for `com.github.ben-manes.caffeine:caffeine:2.9.3`.
- According to my test at https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/2f29acb8845ddde5c3fc7d3d6072dfd8f5574f48 and https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/15ef51ecf88beb819e3844c634b35b115da37a52,  the test results of the APIs of `2.9.3` and `3.1.2` are quite inconsistent.
- This PR is related to the Error detected at https://github.com/apache/shardingsphere/issues/21347 .


## Code sections where the PR accesses files, network, docker or some external service

- (example link to code section)

- Null.

## Checklist before merging
- [x] I have considered including reachability metadata directly in the library or the framework (see [our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md))
- [x] I am the original author of all content provided in the pull request, and I did not copy the content from any other source (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#tests))
- [x] For all tests where I am not the sole author, I have added a comment that proves I may publish them under the specified license (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#tests))
- [x] I have properly formatted metadata files (see [our formatting guide](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#format-metadata-files))
- [x] I have added thorough tests (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#Tests))
- [x] I have filled all places where my pull request accesses files, network, docker, or any other external service (see the section above)

linghengqian reviewed (commented):
- An useless JSON was removed at https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/f1ec6a2c8893c023f7c5564e908564ad1d79b051 .  However, if we remove the following JSON entries involving `java.**`, it will cause nativeTest to hang, so it must be retained.  I haven't seen room for improvement yet.

```json
{
    "condition":{"typeReachable":"com.github.benmanes.caffeine.cache.StripedBuffer"},
    "name":"java.lang.Thread",
    "fields":[{"name":"threadLocalRandomProbe"}]
  }
```

dnestoro commented:
I will review this PR in the coming weeks. 
Since we had issues on [previous PRs](https://github.com/oracle/graalvm-reachability-metadata/pull/224#discussion_r1222753674) @wirthi and @olpaw should also have a look at this one.

wirthi commented:
> I will review this PR in the coming weeks. Since we had issues on [previous PRs](https://github.com/oracle/graalvm-reachability-metadata/pull/224#discussion_r1222753674) @wirthi and @olpaw should also have a look at this one.

I am a bit concerned about the `WriteBehindCacheWriter` class https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-3e57a3d8b35bdd4b113214b87578827af5752e0f36a4a5d7e00ffffa8e8adc44 - This seems to be a direct copy of https://github.com/ben-manes/caffeine/blob/v2.9.3/examples/write-behind-rxjava/src/main/java/com/github/benmanes/caffeine/examples/writebehind/rxjava/WriteBehindCacheWriter.java published under Apache license there.

linghengqian commented:
> > I will review this PR in the coming weeks. Since we had issues on [previous PRs](https://github.com/oracle/graalvm-reachability-metadata/pull/224#discussion_r1222753674) @wirthi and @olpaw should also have a look at this one.
> 
> I am a bit concerned about the `WriteBehindCacheWriter` class https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-3e57a3d8b35bdd4b113214b87578827af5752e0f36a4a5d7e00ffffa8e8adc44 - This seems to be a direct copy of https://github.com/ben-manes/caffeine/blob/v2.9.3/examples/write-behind-rxjava/src/main/java/com/github/benmanes/caffeine/examples/writebehind/rxjava/WriteBehindCacheWriter.java published under Apache license there.

- This is actually a tool class that is easily deduced from the documentation, see https://github.com/ben-manes/caffeine/wiki/Compute . Scenarios outside of RxJava are generally not thought of.
> A write-back extension might implement some or all of the following features:
> Batching and coalescing the operations
> Delaying the operations until a time window
> Performing a batch prior to a periodic flush if it exceeds a threshold size
> Loading from the write-behind buffer if the operations have not yet been flushed
> Handling retrials, rate limiting, and concurrency depending on the the characteristics of the external resource
> See write-behind-rxjava for a simple example using RxJava.

- This PR uses `io.reactivex.rxjava3:rxjava:3.1.6` instead of the corresponding `io.reactivex.rxjava2:rxjava:2.1.12` in https://github.com/ben-manes/caffeine/blob/v2.9.3/examples/write-behind-rxjava/src/main/java/com/github/benmanes/caffeine/examples/writebehind/rxjava/WriteBehindCacheWriter.java . They are not equal. 
- Do you think the unit tests here still need to be removed?

wirthi commented:
> * This is actually a tool class that is easily deduced from the documentation, see https://github.com/ben-manes/caffeine/wiki/Compute . Scenarios outside of RxJava are generally not thought of.

Thanks for the explanation. An _argument_ might be that this is trivial code that anyone would write like this (using the documentation quoted by you). However, this was ~30 LOC, so this argument legally will be on very thin ice. A judge would maybe see that you copied the file from somewhere, removed the comments, and changed the license. We could not accept a contribution like that.

> * The unit tests corresponding to Compute have been removed.

Ok, thanks. In the remaining code, I see no problem any more. I assume all the remaining tests were written by yourself?

linghengqian commented:
> > * This is actually a tool class that is easily deduced from the documentation, see https://github.com/ben-manes/caffeine/wiki/Compute . Scenarios outside of RxJava are generally not thought of.
> 
> Thanks for the explanation. An _argument_ might be that this is trivial code that anyone would write like this (using the documentation quoted by you). However, this was ~30 LOC, so this argument legally will be on very thin ice. A judge would maybe see that you copied the file from somewhere, removed the comments, and changed the license. We could not accept a contribution like that.
> 
> > * The unit tests corresponding to Compute have been removed.
> 
> Ok, thanks. In the remaining code, I see no problem any more. I assume all the remaining tests were written by yourself?

- This PR involves unit tests from my extraction and organization. Refer to https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/f70e8bc7666b7990f2225642bf65e7908cf4510c#diff-e7d1df473b724c8c5c457302a50f2461abdaea72e1e53911502f8c9d6de01338 . That git was originally used to test the use of Graalvm Native Image under WSL2. I don't think the LICENSE issue involved in this PR requires additional handling in another git.

wirthi commented:
It is fine if you already contributed the code somewhere else; if it is YOUR code, that YOU wrote, you can also contribute it here. That's what I assume is the case here. If you can confirm that, I think this is resolved.

@dnestoro will still check the functionality of the PR as usual

Thanks for improving this.

dnestoro commented on metadata/com.github.ben-manes.caffeine/caffeine/2.9.3/reflect-config.json:
```diff
@@ -0,0 +1,203 @@
+},
+{
+  "condition":{"typeReachable":"com.github.benmanes.caffeine.cache.BoundedLocalCache$BoundedLocalLoadingCache"},
+  "name":"com.github.benmanes.caffeine.cache.SSW",
```
Can you explain me what are SSMSR, SSMSW, SSMW, SSSMS, SSMW, SSW...?

linghengqian replied:
- You can think of this as Caffeine's changes to Caffeine's internal class operations based on Quarkus, with more context on Quarkus's requirements. Reference https://github.com/quarkusio/quarkus/issues/10420#issuecomment-654483795 .
- These classes are renamed to their current names after compilation.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/CaffeineTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    void testStatistics() {
+        Cache<String, String> graphs = Caffeine.newBuilder().maximumSize(10_000).recordStats().build();
```
Why are you using `recordStats` here? When are you queering it?  Also, which metadata corresponds to this test? I am asking this because it looks like this test tests same thing as some other tests, and therefore maybe can be removed. Is that correct?

linghengqian replied:
- This seems to be a wrong local refactoring. When verifying the different performances of Guava Cache and Caffeine under Native Image, I reused a set of temporary logic.
- `testStatistics()` has added additional assertions.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/CaffeineTest.java:
```diff
@@ -0,0 +1,74 @@
+        LoadingCache<String, String> firstGraphs = Caffeine.newBuilder().scheduler(Scheduler.systemScheduler()).expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(key -> key.equals("Hello") ? "World" : "Universe");
+        firstGraphs.put("Hello", "World");
+        assertThat(firstGraphs.getIfPresent("Hello")).isEqualTo("World");
```
Is there a need for the `firstGraphs` here? It looks like `cleaner` only operates on `secondGraphs` instance ([here](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-220cd86ae77195bc08dc7249346abd6ccbbbd4ac5f7b828f0bd266b128cc25baR56))

linghengqian replied:
- I looked through the relevant unit tests in my personal git design, and I realized that this involved a bug in the Tracing Agent of the old version of GraalVM CE For JDK11, which collected incorrect JSON entries and led to a refactoring. 
- I refactored `testCleanup()` on GraalVM CE For JDK17.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/EvictionTest.java:
```diff
@@ -0,0 +1,62 @@
+    @Test
+    void testSizeBased() {
+        LoadingCache<String, String> firstGraphs = Caffeine.newBuilder().maximumSize(10_000).build(key -> key.equals("Hello") ? "World" : "Universe");
+        assertThat(firstGraphs.get("Hello")).isEqualTo("World");
```
Isn't this `firstGraphs` testing same as in [testRefresh](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-220cd86ae77195bc08dc7249346abd6ccbbbd4ac5f7b828f0bd266b128cc25baR27) test? The only difference is additional `refreshAfterWrite` function call in the first test. Therefore this `firstGraphs` part of this test can be removed. Is that correct?

linghengqian replied:
- There is a problem with the design of the `testRefresh()` function. The relevant issue record is not in a git, but in my personal blog. When writing relevant unit tests, the unit test for the Refresh function used Mockito to mock `java.util.concurrent.Executor`, and the relevant unit tests failed under GraalVM Native Image due to the use of Mockito's inline plugin. Therefore, when a PR was refactored, this unit test removed the use of the Mockito Java API, making the related unit tests look strange.
- I refactored the `testRefresh()` function to accurately assert actual changes in the Refresh functionality. Apparently this also introduces some extra JSON for `com.github.benmanes.caffeine.cache.SSR`.
- Not sure what I'm missing, but it looks like the version of gradle wrapper currently used by git doesn't support graalvm ce for jdk21.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/PopulationTest.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Test
+    void testLoading() {
+        LoadingCache<String, String> cache = Caffeine.newBuilder().maximumSize(10_000).expireAfterWrite(10, TimeUnit.MINUTES)
```
The only difference (that I can spot) between this test and [testRefresh](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-220cd86ae77195bc08dc7249346abd6ccbbbd4ac5f7b828f0bd266b128cc25baR26) test is usage of `expireAfterWrite` instead of `refreshAfterWrite` function. Besides that, `expireAfterWrite` functionality was exercised in [this test](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-ce2f726eecc1294577266412d9aa4abe59eea73288888fa23df5e9daf6d2ed5cR21), so I believe this test doesn't test anything new and therefore can be removed, right?

linghengqian replied:
- `testLoading()` is now renamed to `testExpireAfterWrite()` and uses logical assertions.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/PopulationTest.java:
```diff
@@ -0,0 +1,68 @@
+    @Test
+    void testAsynchronouslyLoading() throws ExecutionException, InterruptedException {
+        AsyncLoadingCache<String, String> firstCache = Caffeine.newBuilder().maximumSize(10_000).expireAfterWrite(10, TimeUnit.MINUTES)
+                .buildAsync((key, executor) ->
```
Why `executor` arg is needed here when it is not used? Could you use something like `.buildAsync(key -> key.equals("Hello") ? CompletableFuture.supplyAsync(() -> "World") : CompletableFuture.supplyAsync(() -> "Universe"));`? Looks prettier when there are no unused args

linghengqian replied:
- I believe this is impossible. This place actually builds an anonymous implementation class of `com.github.benmanes.caffeine.cache.AsyncCacheLoader`. The implemented `asyncLoad()` method obviously requires `java.util.concurrent.Executor` to be passed in.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/RemovalTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertThat(cache.getIfPresent("Hello")).isNull();
+        cache.invalidateAll(List.of("Hi", "Aloha"));
+        Stream.of("Hi", "Aloha").forEach(s -> assertThat(cache.getIfPresent(s)).isNull());
+        cache.invalidateAll();
```
Isn't it that at this point all keys from the cache ("Hello", "Hi", "Aloha") are already invalidated? Therefore even without this line, [following assert](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-50efd929831e90485065afdd094fbc8af2fb25f5987186c44c3c651b62c70f09R33) will always be correct? Maybe adding one more key-value pair to the Map (that won't be invalidated until this line) will test `cache.invalidateAll()` properly.

linghengqian replied:
- A new key-value pair is used. Solve.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/RemovalTest.java:
```diff
@@ -0,0 +1,47 @@
+    @Test
+    void testExplicitRemovals() {
+        Cache<String, String> cache = Caffeine.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).maximumSize(10_000).build();
+        cache.putAll(Map.of("Hello", "World", "Hi", "Universe", "Aloha", "Universe"));
```
Maybe this `Map.of` creation of a Map can be extracted into a variable, since it is used on two places (at this line and [here](https://github.com/oracle/graalvm-reachability-metadata/pull/388/files#diff-50efd929831e90485065afdd094fbc8af2fb25f5987186c44c3c651b62c70f09R27))?

linghengqian replied:
- Map extracted, marked resolved.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/RemovalTest.java:
```diff
@@ -0,0 +1,47 @@
+        graphs.put("Hello", "World");
+        assertThat(graphs.getIfPresent("Hello")).isEqualTo("World");
+        graphs.invalidate("Hello");
+        assertThat(graphs.getIfPresent("Hello")).isNull();
```
How we can be sure that listeners are actually removed here?

dnestoro commented:
@linghengqian in general the PR looks fine. For future PRs make sure that you only ever add the minimal amount tests needed to verify the metadata your PR adds.

The use of tests in the metadata repository is **not** to cover all possible features of the library but to exercise the code paths in the library that wouldn't work without the metadata your PR adds.

olpaw reviewed (approved):
LGTM, but please address all issues/questions raised by @dnestoro before we can merge.

linghengqian reviewed (commented):
- All reviews have been processed.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/CaffeineTest.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    void testSpecification() {
+        CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m, recordStats");
```
Is `recordStats` a leftover here? Again, it looks like no stats are used in this test. Since you exercised stats in previous test, I guess we can remove it?

linghengqian replied:
- I remembered an issue involved in this unit test based on the comments I wrote for the related unit test a year ago. NativeTest fails when using `CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m");`. But NativeTest works fine when using `CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m, recordStats");`.
- I have now revisited the source code of `com.github.benmanes.caffeine.cache.CaffeineSpec` and I realized that there is a missing JSON entry. This is not automatically collected by Tracing Agent.
```json
{
   "condition":{"typeReachable":"com.github.benmanes.caffeine.cache.NodeFactory"},
   "name":"com.github.benmanes.caffeine.cache.SSMWW",
   "methods":[{"name":"<init>","parameterTypes":["com.github.benmanes.caffeine.cache.Caffeine","com.github.benmanes.caffeine.cache.CacheLoader"," boolean"] }]
}
```
- I removed `CaffeineSpec`'s use of `recordStats` and added the relevant JSON.

dnestoro commented on tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/build.gradle:
```diff
@@ -0,0 +1,35 @@
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+        metadataCopy {
```
Is metadataCopy task needed here at all?

linghengqian replied:
- There is no problem in deleting the corresponding task and `user-code-filter.json`. Just based on the previous discussion, it seems that the Spring team is encouraged to retain the relevant content of generated metadata. I can't find the URL of the original discussion, and you can see the impact of this type of discussion in PRs like #98 .
- I'm assuming the usage of the following shell command must be set up here, unless I'm missing some milestone stuff for GraalVM Native Build Tools. I even suspect that I missed some PRs of the current git, because it seems that my local git is still using gradle wrapper 7.6 which does not support graalce for jdk21.
```shell
sdk use java 17.0.9-graalce
git clone git@github.com:oracle/graalvm-reachability-metadata.git
cd ./graalvm-reachability-metadata/
cd ./tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/
./gradlew -Pagent clean test
./gradlew metadataCopy --task test
./gradlew checkstyle
```
- @dnestoro Do you think these files need to be deleted or retained?

dnestoro replied:
Okay, no problem, we can keep current state of the PR. I think it is ready to be merged
